Add description of multi-seller getInterestGroupAdAuctionData() to B&A explainer (#1403)
WICG/turtledove: FLEDGE_browser_bidding_and_auction_API.md

## Device Orchestrated Multi-Seller Auctions

Bidding and Auction Services auctions can also be used as component auctions. It
is possible for multiple Bidding and Auction Services auctions to be used as
component auctions for an on-device top-level auction.

> For efficiency, a top-level seller can construct encrypted requests for multiple

Reviewer: Will `sellers` use be detectable or must one catch the exception?

```
Uncaught TypeError: Failed to execute 'getInterestGroupAdAuctionData' on 'Navigator': Failed to read the 'seller' property from 'AdAuctionDataConfig': Required member is undefined.
```

Author: Yeah, I think catching the exception is probably the best way to handle it.

component sellers in a single call to `navigator.getInterestGroupAdAuctionData()`:

```javascript
const results = await navigator.getInterestGroupAdAuctionData({
  'sellers': [
    {
      'seller': 'https://www.component-ssp1.example.com',
      'coordinatorOrigin': 'https://publickeyservice.pa.gcp.privacysandboxservices.com/',
    },
    {
      'seller': 'https://www.component-ssp2.example.com',
      'coordinatorOrigin': 'https://publickeyservice.pa.gcp.privacysandboxservices.com/',
  }],
  // 'requestSize' the affects the size of the returned request (optional).
  'requestSize': 51200,
  // 'perBuyerConfig' specifies per-buyer options for size optimizations when
  // constructing the blob (optional).
  'perBuyerConfig': {
    'https://buyer1.origin.example.com': {
      // 'targetSize' specifies the size of the blob to devote to this buyer
      // (optional).
      'targetSize': 8192,
    },
    'https://buyer2.origin.example.com': {}
  }
});

> const requestId = result.requestId;

Reviewer: Still using result.

Note that Chrome balks at syntax here:

```
const requests = {results.requests[0].seller: results.requests[0].request,
                  results.requests[1].seller: results.requests[1].request};
```

FWIW, here's what I replaced it with when testing:

```
            const requestId = results.requestId;
            const sellerRequests = {};
            for( const sellerResult of results.requests) {
                const sellerOrigin = sellerResult.seller;
                sellerRequests[sellerOrigin] = {request, error} = sellerResult;
                if(sellerResult.error) {
                    console.log("Seller '%s' auctionData error: %s", sellerOrigin, sellerResult.error);
                }
            }
```

Author: Thanks. I fixed the example.

## Device Orchestrated Multi-Seller Auctions

Bidding and Auction Services auctions can also be used as component auctions. It
is possible for multiple Bidding and Auction Services auctions to be used as
component auctions for an on-device top-level auction.

For efficiency, a top-level seller can construct encrypted requests for multiple

> component sellers in a single call to `navigator.getInterestGroupAdAuctionData()`:

Reviewer: Some error(s) impact the number of seller resullts one can obtain. 
Specifying three 'clean' sellers and their coordinators produces three result objects.

```
            // test obtaining auctionData blob for multiple sellers
            const result = await navigator.getInterestGroupAdAuctionData({
                // 'seller' works the same as for runAdAuction.
                'sellers': [
                    {
                        // result.error: API not allowed for this origin 
                        // presumably because it is not attested
                        'seller': 'https://www.component-ssp1.example.com',
                        'coordinatorOrigin': 'https://publickeyservice.pa.aws.privacysandboxservices.com/',
                    },
                    {
                        // result.error: Invalid Coordinator
                        'seller': 'https://securepubads.g.doubleclick.net',
                        'coordinatorOrigin': 'https://publickeyservice.pa.typofail.privacysandboxservices.com/',
                    },
                    {
                        // this seller does not have an entry in the API result 
                        'seller': 'https://td.dabbs.net',
                        'coordinatorOrigin': 'https://publickeyservice.pa.gcp.privacysandboxservices.com/',
                    }
                ],
                ....
```

Author: Ah, that's a bug. I filed https://crbug.com/396714215 to track it.

> For efficiency, a top-level seller can construct encrypted requests for multiple
> component sellers in a single call to `navigator.getInterestGroupAdAuctionData()`:

Reviewer 2: ```suggestion
When multiple component sellers desire encrypted requests, they can create them
efficiently in a single call to `navigator.getInterestGroupAdAuctionData()`:
```

```javascript

> const result = await navigator.getInterestGroupAdAuctionData({

Reviewer: Sample code nit: either `result` or `results`, here and below.

Author: Good catch! Thanks.

>   // ‘seller’ works the same as for runAdAuction.

Reviewer: Remove?

Author: Fixed.

>   // ‘seller’ works the same as for runAdAuction.

Reviewer: Note that the API will throw if seller origins are not unique

TypeError: Failed to execute 'getInterestGroupAdAuctionData' on 'Navigator': Each seller in sellers for AdAuctionDataConfig must be unique.
at https://www.mytest.com/demo.html:44:44

  'sellers': [
    {
      'seller': 'https://www.component-ssp1.example.com',
      'coordinatorOrigin': 'https://publickeyservice.pa.gcp.privacysandboxservices.com/',
    },
    {

>       'seller': 'https://www.component-ssp1.example.com',

Reviewer: You probably meant to illustrate different component seller origins...

Author: Thanks. Fixed.

      'coordinatorOrigin': 'https://publickeyservice.pa.gcp.privacysandboxservices.com/',
  }],
  // 'requestSize' the affects the size of the returned request (optional).

>   'requestSize': 51200,

Reviewer: This setup requires a top-level seller driven one-size-fits-all requestSize across all component sellers.

Author: Yes, that's correct. Changing these other parameters has a higher performance impact as there is less that the browser can reuse.
